jetty-server/src/test/java/org/eclipse/jetty/server/HttpOutputInterceptorTest.java

```diff
@@ -0,0 +1,441 @@
+            {
+                baseRequest.setHandled(true);
+                byte[] content = "content to be written".getBytes();
+                HttpOutput httpOutput = baseRequest.getResponse().getHttpOutput();
+                httpOutput.setBufferSize(content.length * 2);
+                httpOutput.write(content);
+
+                try
+                {
+                    // The will close will write the output buffered in HttpOutput to the interceptor which will throw.
+                    httpOutput.close();
+                }
+                catch (Throwable t)
+                {
+                    errorFuture.complete(t);
+                    throw t;
+                }
+            }
+        });
+
+        _server.start();
+        String rawResponse = _localConnector.getResponse("GET /include/path HTTP/1.1\r\nHost: localhost\r\n\r\n");
+        HttpTester.Response response = HttpTester.parseResponse(rawResponse);
+        assertThat(response.getStatus(), is(500));
+
+        // We failed because of the next interceptor.
+        Throwable error = errorFuture.get(5, TimeUnit.SECONDS);
+        assertThat(error.getMessage(), containsString("thrown from interceptor"));
+    }
+
+    @Test
+    public void testAsyncWriteFailed() throws Exception
+    {
+        addInterceptor(nextInterceptor -> new HttpOutput.Interceptor()
+        {
+            @Override
+            public void write(ByteBuffer content, boolean last, Callback callback)
+            {
+                callback.failed(new RuntimeException("callback failed from interceptor"));
+            }
+
+            @Override
+            public HttpOutput.Interceptor getNextInterceptor()
+            {
+                return nextInterceptor;
+            }
+        });
+
+        CompletableFuture<Throwable> errorFuture = new CompletableFuture<>();
+        AtomicInteger onErrorCount = new AtomicInteger(0);
+        _handlerCollection.addHandler(new AbstractHandler()
+        {
+            @Override
+            public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException
+            {
+                baseRequest.setHandled(true);
+                byte[] content = "content to be written".getBytes();
+                AsyncContext asyncContext = request.startAsync();
+                ServletOutputStream outputStream = response.getOutputStream();
+                baseRequest.getResponse().getHttpOutput().setBufferSize(0);
+                outputStream.setWriteListener(new WriteListener()
+                {
+                    @Override
+                    public void onWritePossible() throws IOException
+                    {
+                        while (outputStream.isReady())
+                        {
+                            outputStream.write(content);
```

Comment:
Can you add another variation of this that does the write and then calls `flush` so that we can verify we get the same behaviour if there is an exception in `flush`